src/main/java/taxi/filter/AuthenticationFilter.java

```diff
@@ -0,0 +1,42 @@
+package taxi.filter;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.annotation.WebFilter;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+
+@WebFilter(filterName = "AuthenticationFilter", value = "/*")
+public class AuthenticationFilter implements Filter {
+    private final Set<String> allowedUrls = new HashSet<>();
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+        allowedUrls.add("/login");
+        allowedUrls.add("/drivers/add");
+    }
+
+    @Override
+    public void doFilter(ServletRequest servletRequest,
+                         ServletResponse servletResponse,
+                         FilterChain filterChain)
+            throws IOException, ServletException {
+        HttpServletRequest req = (HttpServletRequest) servletRequest;
+        HttpServletResponse resp = (HttpServletResponse) servletResponse;
+        HttpSession session = req.getSession();
+        Long driverId = (Long) session.getAttribute("driver_id");
+        if (!(driverId == null && allowedUrls.contains(req.getServletPath()))) {
+            resp.sendRedirect("/login");
+            return;
+        }
+        filterChain.doFilter(servletRequest, servletResponse);

```

Comment:
Please do not use ! on combined conditions. It's extremely hard to read and understand such structure.
If you remove ! you just need to switch code in if and else to get the same:
```suggestion
        if (driverId == null && allowedUrls.contains(req.getServletPath())) {
            filterChain.doFilter(servletRequest, servletResponse);
            return;
        }
        resp.sendRedirect(request.getContextPath() + "/login");
```

And now we see that this condition doesn't look like correct as we need following:
```suggestion
        if (driverId != null || allowedUrls.contains(req.getServletPath())) {
            filterChain.doFilter(servletRequest, servletResponse);
            return;
        }
        resp.sendRedirect(request.getContextPath() + "/login");
```
or
```suggestion
        if (driverId == null && !allowedUrls.contains(req.getServletPath())) {
            resp.sendRedirect(request.getContextPath() + "/login");
            return;
        }
        filterChain.doFilter(servletRequest, servletResponse);
```